core/trino-main/src/test/java/io/trino/operator/TestHashAggregationOperator.java

```diff
@@ -698,6 +713,77 @@ public void testMemoryTracking()
         assertEquals(driverContext.getRevocableMemoryUsage(), 0);
     }
 
+    @Test
+    public void testAdaptivePartialAggregation()
+            throws Exception
+    {
+        List<Integer> hashChannels = Ints.asList(0);
+
+        HashAggregationOperatorFactory operatorFactory = new HashAggregationOperatorFactory(
+                0,
+                new PlanNodeId("test"),
+                ImmutableList.of(BIGINT),
+                hashChannels,
+                ImmutableList.of(),
+                PARTIAL,
+                ImmutableList.of(LONG_MIN.createAggregatorFactory(PARTIAL, ImmutableList.of(0), OptionalInt.empty())),
+                Optional.empty(),
+                Optional.empty(),
+                100,
+                Optional.of(DataSize.ofBytes(1)), // this setting makes operator to flush after each page
+                joinCompiler,
+                blockTypeOperators,
+                // use 5 rows threshold to trigger adaptive partial aggregation after each page flush
+                new PartialAggregationControllerFactory(true, 5, 0.8));
+
+        // First operator will trigger adaptive partial aggregation after the first page
+        List<Page> operator1Input = rowPagesBuilder(false, hashChannels, BIGINT)
+                .addSequencePage(10, 0) // first page will be hashed but the values are unique anyway, so it will trigger adaptation
+                .addBlocksPage(createRLEBlock(1, 10)) // second page would be hashed to existing value 1. but if adaptive PA kicks in, the raw values will be passed on
+                .build();
+        List<Page> operator1Expected = rowPagesBuilder(BIGINT, BIGINT)
+                .addSequencePage(10, 0, 0)
```

Comment:
so this intermediate page has 10 positions? Could we use aggregation and input data that would actually squash input positions (e.g. all rows belong to same group)

Reply:
Ok, so can we use 9 unique rows out of 10 rows? Or maybe you change ration from 0.8 into 0.5? The reason is the I would like to see that aggregation actually happens before it's disabled